pytorch_lightning/callbacks/model_checkpoint.py

```diff
@@ -342,19 +327,16 @@ def on_validation_end(self, trainer, pl_module):
         if trainer.running_sanity_check:
             return
 
-        # TODO: remove when dict results are deprecated
-        self.__warn_deprecated_monitor_key()
-
         metrics = trainer.logger_connector.callback_metrics
         epoch = trainer.current_epoch
 
-        # support structured results
-        if metrics.get("checkpoint_on") is not None:
-            self.monitor = "checkpoint_on"
-
-        # conditioned val metrics override conditioned train loop metrics
-        if metrics.get("val_checkpoint_on") is not None:
-            self.monitor = "val_checkpoint_on"
+        # validate metric
+        if not self._is_valid_monitor_key(metrics):
+            keys = list(metrics.keys())
+            m = f"""
+                ModelCheckpoint(monitor='{self.monitor}') not found in the returned metrics ({keys}),
+                "did you call result.log(f'{self.monitor}', tensor)?"""
+            raise MisconfigurationException(m)
 
         if self.save_top_k == 0:
             # no models are saved
```

Comment:
this check should be to right after line 327 so we can exit earlier

Reply:
there's no real difference though. You need the metrics dict on 330 to know if there's an error

Reply:
i guess it's not clear on the UI. I meant moving the `if self.save_top_k == 0` check up to even before we check metrics. if `save_top_k = 0`, then there's no need to check metrics since there are no checkpoints to save

Reply:
done